src/main/java/org/jooq/lambda/Iterators.java

```diff
@@ -0,0 +1,116 @@
+package org.jooq.lambda;
+
+import java.util.Iterator;
+import java.util.function.Predicate;
+
+/**
+ * Created by billoneil on 7/26/17.
+ */
+public class Iterators {
+    private enum PredicateCheckState { NOT_TESTED, MORE, ONE_MORE, NO_MORE }
+
+    static <E> PeekingIterator<E> peeking(Iterator<E> iterator) {
+        return new PeekingIteratorImpl<E>(iterator);
+    }
+
+    static <E> Iterator<E> takeWhileIterator(Iterator<E> iterator, Predicate<E> predicate) {
+        return takeWhileIterator(peeking(iterator), predicate);
+    }
+
+    static <E> Iterator<E> takeWhileIterator(PeekingIterator<E> iterator, Predicate<E> predicate) {
```

Comment:
I meant to say: Perhaps `Seq.take` should use this new iterator. In that case, we should open up several tickets and do things one step at a time...